engine/gamesys/src/gamesys/scripts/script_sys_gamesys.cpp

```diff
@@ -0,0 +1,476 @@
+        {
+            dmMutex::ScopedLock lk(g_SysModule.m_LoadRequestsMutex);
+
+            // We currently don't do anything about duplicated requests here (should we?)
+            for (int i = 0; i < g_SysModule.m_LoadRequests.Capacity(); ++i)
+            {
+                LuaRequest* active_request = g_SysModule.m_LoadRequests.GetByIndex(i);
+                if (active_request && active_request->m_PathHash == path_hash)
+                {
+                    dmLogWarning("sys.load_buffer_async called with path '%s' that is already pending", path);
+                    break;
+                }
+            }
+
+            if (g_SysModule.m_LoadRequests.Full())
+            {
+                g_SysModule.m_LoadRequests.Allocate(4);
+            }
+
+            LuaRequest* request     = new LuaRequest();
+            request->m_Path         = path;
+            request->m_PathHash     = path_hash;
+            request->m_CallbackInfo = callback_info;
+            request->m_Status       = REQUEST_STATUS_INITALIZED;
+            request->m_Handle       = g_SysModule.m_LoadRequests.Put(request);
+            request->m_Payload      = 0;
+            DispatchRequest(request);
+            lua_pushnumber(L, request->m_Handle);
+        }
+
+        assert(top + 1 == lua_gettop(L));
+        return 1;
+    }
+
+    static const luaL_reg ScriptImage_methods[] =
+    {
+        {"load_buffer",       Sys_LoadBuffer},
+        {"load_buffer_async", Sys_LoadBufferAsync},
+        {0, 0}
+    };
+
+    /*# an asyncronous request has finished successfully
+     * @name sys.REQUEST_STATUS_FINISHED
+     * @variable
+     */
+
+    /*# an asyncronous request is unable to read the resource
+     * @name sys.REQUEST_STATUS_ERROR_IO_ERROR
+     * @variable
+     */
+
+    /*# an asyncronous request is unable to locate the resource
+     * @name sys.REQUEST_STATUS_ERROR_NOT_FOUND
+     * @variable
+     */
+
+    void ScriptSysGameSysRegister(const ScriptLibContext& context)
+    {
+        lua_State* L = context.m_LuaState;
+        int top = lua_gettop(L);
+        (void)top;
+
+        luaL_register(L, "sys", ScriptImage_methods);
+
+    #define SETCONSTANT(name, val) \
+        lua_pushnumber(L, (lua_Number) val); \
+        lua_setfield(L, -2, #name);\
+
+        SETCONSTANT(REQUEST_STATUS_FINISHED,        REQUEST_STATUS_FINISHED);
+        SETCONSTANT(REQUEST_STATUS_ERROR_IO_ERROR,  REQUEST_STATUS_ERROR_IO_ERROR);
+        SETCONSTANT(REQUEST_STATUS_ERROR_NOT_FOUND, REQUEST_STATUS_ERROR_NOT_FOUND);
+    #undef SETCONSTANT
+
+        lua_pop(L, 1);
+        assert(top == lua_gettop(L));
+
+        g_SysModule.m_Factory           = context.m_Factory;
+        g_SysModule.m_LoadQueue         = dmLoadQueue::CreateQueue(context.m_Factory);
+        g_SysModule.m_LoadRequestsMutex = dmMutex::New();
+    }
+
+    bool ScriptSysGameSysUpdate(const ScriptLibContext& context)
+    {
+        dmMutex::ScopedLock lk(g_SysModule.m_LoadRequestsMutex);
```

Comment:
I wonder if we can get stalls here if the load function has taken the mutex any doing a massive load from a http request or some such scenario? In the update we need to look at the requests in the list of requests, but the load function will touch the data, so we need to solve this somehow.

Reply:
You can use `dmMutex::TryLock()` if you want?